### PR TITLE
Krhubert/api core

### PR DESCRIPTION
**Pull request by NicolasMahe**

### CODE REVIEW

File: api/core/get_service.go [hunk 1]
```diff
@@ -2,17 +2,17 @@ package core
...
-func (s *Server) GetService(ctx context.Context, request *GetServiceRequest) (reply *GetServiceReply, err error) {
+// GetService fetches a service from the database and returns it.
+func (s *Server) GetService(ctx context.Context, request *GetServiceRequest) (*GetServiceReply, error) {
 	service, err := services.Get(request.ServiceID)
 	if err != nil {
-		return
+		return nil, err
 	}
-	reply = &GetServiceReply{
+	return &GetServiceReply{
```
Comment (antho1404) [inline api/core/get_service.go]: can compact in only one `return &GetServiceReply{Service: &service}, err`

File: api/core/list_services.go [hunk 2]
```diff
@@ -2,17 +2,17 @@ package core
...
-func (s *Server) ListServices(ctx context.Context, request *ListServicesRequest) (reply *ListServicesReply, err error) {
+// ListServices returns all services from the database.
+func (s *Server) ListServices(ctx context.Context, request *ListServicesRequest) (*ListServicesReply, error) {
 	services, err := services.All()
 	if err != nil {
-		return
+		return nil, err
 	}
-	reply = &ListServicesReply{
+	return &ListServicesReply{
```
Comment (antho1404) [inline api/core/list_services.go]: can also compact `return &ListenServicesReply{Services: services}, err`

File: api/core/listen_result.go [hunk 3]
```diff
@@ -37,39 +37,37 @@ func (s *Server) ListenResult(request *ListenResultRequest, stream Core_ListenRe
...
-func validateOutputKey(service *service.Service, taskKey string, outputFilter string) (err error) {
+func validateOutputKey(service *service.Service, taskKey string, outputFilter string) error {
 	if outputFilter == "" || outputFilter == "*" {
-		return
+		return nil
 	}
 	if taskKey == "" {
-		err = errors.New("Cannot filter output without specifying a task")
-		return
+		return errors.New("Cannot filter output without specifying a task")
```
Comment (antho1404) [inline api/core/listen_result.go]: let's use `fmt.Errorf` here to like that we have the same way to return errors (and later even maybe some custom error type)

File: api/core/stop.go [hunk 4]
```diff
@@ -2,16 +2,18 @@ package core
...
-		return
+		return nil, err
+	}
+	if err := service.Stop(); err != nil {
+		return nil, err
 	}
-	err = service.Stop()
-	reply = &StopServiceReply{}
-	return
+	return &StopServiceReply{}, nil
```
Comment (antho1404) [inline api/core/stop.go]: can compact with `return &StopServiceReply{}, service.Stop()`

**Comment by MarinX** [comment]
@antho1404 updated with the compact fixes.

**Comment by NicolasMahe** [comment]
@MarinX why did you edit files from the `api/client` package? There is another PR dedicated for `api/client`.

**Comment by antho1404** [comment]
let's merge the #333 and see what's left to merge on this one

File: api/core/listen_result.go [hunk 5]
```diff
@@ -37,39 +36,37 @@ func (s *Server) ListenResult(request *ListenResultRequest, stream Core_ListenRe
...
-		err = errors.New("Task '" + taskKey + "' doesn't exist in this service")
-		return
+		return fmt.Errorf("Task %q doesn't exist in this service", taskKey)
 	}
+
+	var err error
 	_, ok = task.Outputs[outputFilter]
 	if !ok {
-		err = errors.New("Output '" + outputFilter + "' doesn't exist in the task '" + taskKey + "' of this service")
+		err = fmt.Errorf("Output %q doesn't exist in the task %q of this service", outputFilter, taskKey)
```
Comment (ilgooz) [inline api/core/listen_result.go]: directly return here

File: api/core/listen_result.go [hunk 6]
```diff
@@ -37,39 +36,37 @@ func (s *Server) ListenResult(request *ListenResultRequest, stream Core_ListenRe
...
 	}
-	return
+	return err
```
Comment (ilgooz) [inline api/core/listen_result.go]: return nil

File: api/core/listen_result.go [hunk 7]
```diff
@@ -37,39 +36,37 @@ func (s *Server) ListenResult(request *ListenResultRequest, stream Core_ListenRe
...
+		return fmt.Errorf("Cannot filter output without specifying a task")
 	}
 	task, ok := service.Tasks[taskKey]
 	if !ok {
-		err = errors.New("Task '" + taskKey + "' doesn't exist in this service")
-		return
+		return fmt.Errorf("Task %q doesn't exist in this service", taskKey)
 	}
+
+	var err error
```
Comment (ilgooz) [inline api/core/listen_result.go]: rm this

File: api/core/start.go [hunk 8]
```diff
@@ -2,16 +2,18 @@ package core
...
-// StartService fetch a service in the db and starts it
-func (s *Server) StartService(ctx context.Context, request *StartServiceRequest) (reply *StartServiceReply, err error) {
+// StartService fetches a service from the database and starts it.
+func (s *Server) StartService(ctx context.Context, request *StartServiceRequest) (*StartServiceReply, error) {
 	service, err := services.Get(request.ServiceID)
 	if err != nil {
-		return
+		return nil, err
+	}
+	if _, err = service.Start(); err != nil {
```
Comment (ilgooz) [inline api/core/start.go]: replace `service.Start()` with the nil error value at line 18.

File: api/core/listen_result.go (hunk 5 above)
Comment (NicolasMahe) [reply api/core/listen_result.go]: Done

File: api/core/listen_result.go (hunk 6 above)
Comment (NicolasMahe) [reply api/core/listen_result.go]: Done

File: api/core/listen_result.go (hunk 7 above)
Comment (NicolasMahe) [reply api/core/listen_result.go]: Done

File: api/core/start.go (hunk 8 above)
Comment (NicolasMahe) [reply api/core/start.go]: Done

**Pull request merged**